src/cuckoocache.h

```diff
@@ -0,0 +1,450 @@
+ *
+ *
+ * Note on function names:
+ *   - The name "allow_erase" is used because the real discard happens later.
+ *   - The name "please_keep" is used because elements may be erased anyways on insert.
+ *
+ * @tparam Element should be a POD type that is 32-alignable
+ * @tparam Hash should be a function/callable which takes a template parameter
+ * hash_select and an Element and extracts a hash from it. Should return
+ * high-entropy hashes for `Hash h; h<0>(e) and h<1>(e)`.
+ */
+template <typename Element, typename Hash>
+class cache
+{
+    static_assert((sizeof(Element) % 32) == 0, "Invalid Element Size.");
+
+public:
+private:
+    /** table stores all the elements */
+    std::vector<Element> table;
+
+    /** size stores the total available slots in the hash table */
+    uint32_t size;
+
+    /** The bit_packed_atomic_flags array is marked mutable because we want
+     * garbage collection to be allowed to occur from const methods */
+    mutable bit_packed_atomic_flags collection_flags;
+
+    /** epoch_flags tracks how recently an element was inserted into
+     * the cache. true denotes recent, false denotes not-recent. See insert()
+     * method for full semantics.
+     */
+    mutable std::vector<bool> epoch_flags;
+
+    /** epoch_heuristic_counter is used to determine when a epoch
+     * might be aged & an expensive scan should be done.
+     * epoch_heuristic_counter is incremented on insert and reset to the
+     * new number of inserts which would cause the epoch to reach
+     * epoch_size when it reaches zero.
+     */
+    uint32_t epoch_heuristic_counter;
+
+    /** epoch_size is set to be the number of elements supposed to be in a
+     * epoch. When the number of non-erased elements in a epoch
+     * exceeds epoch_size, a new epoch should be started and all
+     * current entries demoted. epoch_size is set to be 45% of size because
+     * we want to keep load around 60%, and we support 3 epochs at once --
+     * one "dead" which has been erased, one "dying" which has been marked to be
+     * erased next, and one "living" which new inserts add to.
+     */
+    uint32_t epoch_size;
+
+    /** depth_limit determines how many elements insert should try to replace.
+     * Should be set to log2(n)*/
+    uint8_t depth_limit;
+
+    /** hash_function is a const instance of the hash function. It cannot be
+     * static or initialized at call time as it may have internal state (such as
+     * a nonce).
+     * */
+    const Hash hash_function;
+
+    /** compute_hashes is convenience for not having to write out this
+     * expression everywhere we use the hash values of an Element.
+     *
+     * @param e the element whose hashes will be returned
+     * @returns std::array<uint32_t, 8> of deterministic hashes derived from e
+     */
+    inline std::array<uint32_t, 8> compute_hashes(const Element& e) const
+    {
+        return {hash_function.template operator()<0>(e) % size,
```

Comment:
Yes, this is a rather slow operation, but I don't quite fully understand your operation... doesn't that only work if you're at power of two size?